src/sql/workbench/contrib/query/browser/messagePanel.ts

```diff
@@ -188,30 +197,54 @@ export class MessagePanel extends Disposable {
 	}
 
 	public set queryRunner(runner: QueryRunner) {
+		this.runner = runner;
+
 		if (this.currenturi) {
 			this._treeStates.set(this.currenturi, this.tree.getViewState());
 		}
 		this.queryRunnerDisposables.clear();
 		this.reset();
 		this.currenturi = runner.uri;
-		this.queryRunnerDisposables.add(runner.onQueryStart(() => this.reset()));
-		this.queryRunnerDisposables.add(runner.onMessage(e => this.onMessage(e)));
-		this.onMessage(runner.messages, true);
+
+		this.queryRunnerDisposables.add(runner.onQueryStart(this.onQueryStart, this));
+		this.queryRunnerDisposables.add(runner.onMessage(this.onMessage, this));
+		this.queryRunnerDisposables.add(runner.onResultSet(this.onResultSet, this));
+		this.queryRunnerDisposables.add(runner.onResultSetUpdate(this.updateResultSet, this));
+
+		if (this.queryResultsWriterStatus.isWritingToGrid()) {
+			this.onMessage(runner.messages, true);
+		}
 	}
 
-	private onMessage(message: IQueryMessage | IQueryMessage[], setInput: boolean = false) {
-		if (isArray(message)) {
-			this.model.messages.push(...message);
-		} else {
-			this.model.messages.push(message);
+	public changeQueryResultsWriter() {
+		if (this.queryResultsWriterStatus.isWritingToGrid()) {
+			this.queryResultsWriter = new MessagesPanelQueryResultsWriter(this.model, this.tree, this._treeStates, this.currenturi);
 		}
-		if (setInput) {
-			this.tree.setInput(this.model, this._treeStates.get(this.currenturi));
-		} else {
-			this.tree.updateChildren();
+		else {
+			this.queryResultsWriter = this.instantiationService.createInstance(FileQueryResultsWriter, this.runner);
 		}
 	}
 
+	private onQueryStart() {
+		this.queryResultsWriter.onQueryStart();
```

Comment:
it may be better to have the result writer subscribe to runner messages directly (assuming the events can be multicast and the order of the methods doesn't matter).  not a big deal, just seems like an extra level of indirection.

Reply:
Is this how we subscribe directly to runner messages or am I misunderstanding?
 `this.queryRunnerDisposables.add(runner.onQueryStart(this.queryResultsWriter.onQueryStart, this));`